### ISSUE TITLE
Better support for monorepos

### ISSUE THREAD

**Issue by bellini666** [jose-elias-alvarez/null-ls.nvim#289]
### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues) and there are no existing ones with the same request.

### Feature description

Hi,

Where I work we use a lot of monorepos, and those doesn't even contain the same language (e.g. in a project the `frontend` directory will be a typescript project while the `backend` will be python)

The way null-ls works has 2 issues for monorepos like those:

1) Some linters/formatters needs to run with their `cwd` set to the directory where the configuration is defined. For example, inside the `backend` folder we have a `.flake8` file which configures flake8, but if null-ls runs it from the root (the directory above, where .git is) it will not use those configs.

I could get around this by passing a custom `cwd` function, but I'm not sure it is the proper way (I'll explain why bellow)

2) The major issue: We need to run the specific installed linters/formatters inside those projects. For example, we have some `flake8` plugins installed in the project which will not be included in the global flake8 binary. Idially the `flake8` installed in the project should be used.

To workaround that I did the something like this:

```lua
null_ls.config({
    sources = {
        null_ls.builtins.diagnostics.flake8.with({
            command = function()
                if ROOT_DIRS.python ~= nil then
                    local match = vim.fn.glob(path.join(ROOT_DIRS.python, ".venv", "bin", "flake8"))
                    if match ~= "" then
                        return match
                    end
                end
                return "flake8"
            end,
            cwd = function(params)
                return ROOT_DIRS.python or params.cwd
            end,
        }),
    }
})
```

What is the problem there? That `ROOT_DIRS.python` is defined like this:

```lua
ROOT_DIRS = {
    python = nil,
    node = nil,
}
nvim_lsp.pyright.setup({
    before_init = function(_, config)
        ROOT_DIRS.python = config.root_dir
    end,
})
```

So, I'm depending that the python LSP runs first because it will find the properly root dir for that specific file in the buffer, so that when null_ls is going to retrieve the command, I already have the python root dir.

I cannot simply set the root dir to something else because each one will have its own root and its own formatters/linters (in my example, I want to run `flake8`, `isort` and `black` iside the `backend` dir and `prettier` and `eslint` inside the `frontend` dir). I did the same for each source, I'm not showing everything here for simplicity =P

In summary, this works and I got everything running. But I'm afraid of that "if null-ls asks for the cmd first?" and also it doesn't seem to be a clean and elegand solution.

Maybe there's a way to improve that kind of workflow?

### Help

No

**Comment by bellini666** [comment]
> I can imagine wanting to use a different command based on the presence of a config file or something else about the project's state that might change while the buffer is loaded

Is this something common? Like, even in my case, I would understand opening a file and `flake8` sticking to the system's binary even after I create a virtualenv inside the main dir, and to get it to use it I would need to "reload" neovim, or at least reload the LSP.  Just like neovim, after I open it, if I modify my `init.lua` file and add a mapping there I don't expect that it will register it unless I reload that source or close and open it again, but there might be some specific case that you have in mind that I'm not familiar with.

Having said that, either going full cache on `before_init`/`on_attach` or giving the possibility of those being changed on each run, if that means that each file can have its own `command`/`cwd` then the problem this issue is bringing to the table is totally solved!

Just to give my some more thoughts about the issue, maybe the caching thing can be decided like this (just some thoughts):

1) Someone uses the builtin directly, without providing a function to command or cwd. Then those can be cached at before_init/on_attach (cwd will basically be the found root_dir) and null-ls don't ever need to ask for them again

2) Someone uses the builtin but asks to `prefer_local` as a list of possibilities. At before_init/on_attach null-ls checks for those and, if any of those match (giving priority to the order of the list) then that bin is used as command and the cwd where it was found is used as the cwd. Those are cached and null-ls don't ever need to ask for them again

Note here that 1 and 2 should, AFAIK, cover most user's use cases, including all that I pointed in this issue. And they should be very performatic since null-ls should not have to ask for them ever again.

Now for the extra cases:

3) If the user provides a `cwd` function it will be used in preference to the root_dir cwd and also the `prefer_local` cwd. null-ls asks for it at before_init/in_attach and don't ever need to ask for it again.

4) Just like for `cwd`, providing a custom `command` would do the same as above for it, caching it too.

5) Now, here comes the part where everything changes: A new option, `cache_options` (or something like that), which could be initially set to true (I believe that most would want the cache) where, if one needs `cwd` or even `command` to be asked every time, than he could set this to `false` and the function passed to one or both `cwd` and/or `command` would get called at all runs.

Obviously the `command` one would need more time to really be implemented, like you said, but the `cwd` one is just flagging it as it already acts like `cache_options=false` when there's a function there.

That `cache` could also be separated into parameters for each one, or even be a table mapping each kind (e.g. `cache = {command: true, cwd: false}`)

Again, just some ideas! Feel free to ignore them. At the end of the day, adding the `prefer_local` and retrieving the command and cwd (when it is not a function provided by the user) for it for each different file/buffer, than this issue is totally solved (with the small gain in performance by caching cwd at before_init/on_attach time). The rest  are improvements that solves other issues and improves the overall solution, but doesn't prevent it from starting as simple as this.

Please tell me if you need any help testing or even with the development of this!

**Comment by YorickPeterse** [comment]
If it's of any use, I have a similar ish case (which overlaps with https://github.com/jose-elias-alvarez/null-ls.nvim/issues/270):

If a Ruby project has a `Gemfile.lock`, I want to run RuboCop with `bundle exec` but _only_ if the `Gemfile.lock` includes RuboCop. If it doesn't include it, I don't want to run RuboCop at all. 

I sort of managed to achieve it like this:

```lua
nls.builtins.diagnostics.rubocop.with({
  command = 'ruby-exec',
  args = vim.list_extend(
    { 'rubocop' },
    nls.builtins.diagnostics.rubocop._opts.args
  ),
  runtime_condition = function(params)
    return not rubocop_with_bundler(params.client_id)
  end,
}),
                                                      
nls.builtins.diagnostics.rubocop.with({
  command = 'ruby-exec',
  args = vim.list_extend(
    { 'bundle', 'exec', 'rubocop' },
    nls.builtins.diagnostics.rubocop._opts.args
  ),
  runtime_condition = function(params)
    return rubocop_with_bundler(params.client_id)
  end,
}),
```

Here `rubocop_with_bundler` is a function that takes the LSP client ID (to get the root directory), then determines if `bundle exec` should be used. The `ruby-exec` command is just a small shell wrapper I use to ensure the correct Ruby version is used. 

Since `runtime_condition` is called every time the linter is invoked, I do make sure the result of the function is cached. This does lead to a bit of boilerplate, but it's manageable.

Probably the most annoying part is having to list the RuboCop linter twice in the sources list, rather than once with a dynamic command and arguments.

**Comment by jose-elias-alvarez** [comment]
@bellini666 Thank you again for the feedback! I don't have a specific case for dynamic commands in mind, but working on this project has taught me that there is an incredible variety of use cases and project structures out there, so I prefer to keep things flexible and leave it to users / plugins to solve language-specific issues. 

For now, I think it's best to keep things simple, and the simplest option I can think of is to have `prefer_local` always cache `command` and `cwd` per buffer on the first run. Fundamentally, this is just a wrapper for functionality that users can already implement, so I don't think we need to handle every possible config option, but I'm always open to improving the user experience and we can definitely keep working on this in the future.

@YorickPeterse Maybe I'm misunderstanding, but would setting `args` dynamically solve the issue?

```lua
nls.builtins.diagnostics.rubocop.with({
    command = "ruby-exec",
    args = function(params)
        return vim.list_extend(
            rubocop_with_bundler(params.client_id) and { "bundle", "exec", "rubocop" } or { "rubocop" },
            nls.builtins.diagnostics.rubocop._opts.args
        )
    end,
})
```

**Comment by bellini666** [comment]
@jose-elias-alvarez hahaha yea, I know what you mean :)

And that option seems like the best one for me too. It solves this issue and also improves the overall performance by caching command/cwd for people that have those defined.

**Comment by YorickPeterse** [comment]
@jose-elias-alvarez Dynamic arguments is something I found out about after writing my comment. It indeed cleans things up. However, it only works because my `command` is a static wrapper. Without that, I'd need a dynamic `command` as well.

**Comment by jose-elias-alvarez** [comment]
Ah, I see now. That case should be handled by `dynamic_command` (or whatever we end up calling it), which will also be added in the process of handling everything else mentioned here.

**Comment by mjlbach** [comment]
FYI lspconfig defaults to cwd == resolved_root for servers now

**Comment by jose-elias-alvarez** [comment]
@bellini666 Could you (and anyone else interested in this) try #349?

**Comment by jose-elias-alvarez** [comment]
@bellini666 Whoops, totally forgot about the `cwd` issue. I just pushed another commit to the PR that should take care of it, I think. Let me know.

**Comment by bellini666** [comment]
Hey @jose-elias-alvarez 

Now it worked! :)

I have a small corner case, which will not affect me, but want to make sure to comment here. Take this as an example:

- root_dir:
  - .git
  - foo/
    - bar.py
  - backend/
    - .venv
      - bin
        - flake8
        - isort
        - black
    - dir1/
      - some_file.py

When I opened `some_file.py` now it worked perfectly! cwd as set to backend and it used flake8 from .venv/bin directory.

Then I opened at another buffer the `bar.py` file. It also worked perfectly as it used my global flake8 and the root_dir as cwd.

When exchanging buffers, each one uses the correct found binary (the global one for bar.py and the local one for some_file), which is correct! But after opening `bar.py`, the `cwd` changed not only for it, but also for from `some_file.py`, so it started to use the wrong cwd too.

Like I said, it is a very specific corner case that doesn't actually affect me, but you might want to workaround that too.

Other than that, everything working really well! :)

**Comment by jose-elias-alvarez** [comment]
@bellini666 Right, I see that we should also index `cwd` by `bufnr` to handle cases like this. I just pushed a commit that should help with that, too. Could you let me know if that takes care of it?

**Comment by bellini666** [comment]
@jose-elias-alvarez when trying the last change I got this error (don't know how to get it as plain text):

![Screenshot from 2021-11-16 15-36-36](https://user-images.githubusercontent.com/134025/142045373-8f11d4b0-a089-4626-b521-993a36452a69.png)

**Comment by jose-elias-alvarez** [comment]
@bellini666 Sorry, there was a typo when returning the cached command. Could you try again?

**Comment by bellini666** [comment]
@jose-elias-alvarez np! :)

Now it is perfect! No remaining issues.

**Comment by bellini666** [comment]
Hey @jose-elias-alvarez ,

Eagerly waiting for this to be released! Anything else that you need to test that I can help with?

**Comment by bellini666** [comment]
@jose-elias-alvarez awesome, it is me that thank you for the attention :)

I'll surely let you know if I run into any issues

**Comment by bellini666** [comment]
@jose-elias-alvarez btw, the change to prefer_local instead of having it running my custom function everytime made everything a lot faster. Like, to the point that I can notice it linting and formatting faster than before. Thank you very much for this!

**Comment by jose-elias-alvarez** [comment]
@bellini666 Awesome, great to hear!

**Comment by nyngwang** [comment]
Hi, I'm using PDM, which my `pyright`, `autopep8`, `flake8` will be installed in the folder `myproject/__pypackages__/`, and I will have to run command like `pdm run autopep8 ...`, i.e. prepending `pdm run`, to use those local packages inside the `__pypackages__/` folder. How should I set this up with `null-ls`?

**Comment by nyngwang** [comment]
> [...] and I will have to prepend commands like `pdm run autopep8 ...`(i.e. with `pdm run ...`) to use these _local packages_.

__good news__: I just resolved this with these settings, e.g.:

```lua
diagnostics.flake8.with({
  command = 'pdm',
  args = { 'run', 'flake8', '--stdin-display-name', '$FILENAME', '-' },
  timeout = 10000
}),
```

**Comment by jose-elias-alvarez** [comment]
@nyngwang By default `vim.lsp.buf.formatting()` will not save the file after it runs, so what you are seeing is expected behavior, regardless of whether you're using `BufWritePre` or `BufWritePost`. You'd either have to override the handler or use something like [this custom function](https://github.com/jose-elias-alvarez/null-ls.nvim/wiki/Async-formatting) to apply results and save.

**Comment by nyngwang** [comment]
@jose-elias-alvarez: I'm using `tsserver` and `eslint`, and their on-save behaviors are very smooth(it looks like there is only one save, but I'm not sure) without additional settings. Is it possible for null-ls to be faster in the future? Or at least the setup could be made more simple.

**Comment by jose-elias-alvarez** [comment]
> @jose-elias-alvarez: I'm using `tsserver` and `eslint`(setup via `nvim-lspconfig`), and their on-save behaviors are very smooth(it looks like there is only one save, but I'm not sure) without additional settings. Is it possible for null-ls to be as fast like them in the future? (Or at least the linked setup could be _somehow_ simplified.)

It's likely that something else in your config is affecting this, because absent other factors, the way you set up formatting in the snippet above should require saving twice if the file is actually formatted (once to trigger `vim.lsp.buf.formatting()` and again to write any changes made). It's not really a speed issue, and I also don't plan on significantly deviating from the LSP formatting API.

Anyways, this is getting away from the original issue, so if we want to continue this discussion I think we should move it elsewhere.
